### DISCUSSION TITLE
TCP-packet processing in outbound proxy in linkerd-proxy

### DISCUSSION THREAD

**Discussion by NikitaSkrynnik** [linkerd/linkerd2#10145] [Help] [answered]
Hello. I'm trying to understand how `linkerd-proxy` works internally. I'm not very familiar with Rust. I wonder where in code TCP-packet processing in `outbound proxy` begins. Can somebody help me, please?

**Comment by kleimkuhler** [comment, marked as answer]
In order to start understanding the proxy, understanding [tower](https://docs.rs/tower/latest/tower/), one of the core libraries it uses, will be very helpful. If you read through those docs, you'll start to get a sense about how different layers and stacks compose to make any of the various "services" that you'll see in the proxy code.

As for the outbound TCP-related code in the proxy, you'll want to start in `linkerd/app/outbound/src/lib.rs`, specifically in [fn mk_proxy](https://github.com/linkerd/linkerd2-proxy/blob/6d2abbcf26f3e487acdf56230dc628a01e67b9dd/linkerd/app/outbound/src/lib.rs#L194-L217). You can see how we start with creating the `logical` and `forward` stacks, and then use those to create a single stack that handles things like discovery and instrumentation.

If you're not familiar with Rust, I'd recommend working in an environment with the Rust language server so that you can click through to function definitions and get an idea about what types are being worked with.

**Reply by NikitaSkrynnik** [reply]
Thank you for the answer. Could you, please, also tell me for what types of requests this [branch in switch_logical](https://github.com/linkerd/linkerd2-proxy/blob/main/linkerd/app/outbound/src/switch_logical.rs#L52) happens? I can't understand how we can provide `logical_addr` in TCP packets coming to outbound proxy.

`logical_addr` consists of dns name and port, doesn't it?

**Reply by kleimkuhler** [reply]
`logical_addr` is part of the destination profile response from the [proxy API](https://github.com/linkerd/linkerd2-proxy-api/blob/main/proto/destination.proto#L148-L190). If we have a `logical_addr`, that means we're dealing with a destination that is named (i.e. a Service). This means that there are multiple endpoints that we can perform load balancing, traffic splitting, etc. on.

Above that switch you can see we do the same check on if it's a single endpoint (`if let Some((addr, metadata)) = rx.endpoint()`). In that case, we can't perform those things I mentioned above which is why we create a different "stack".
